Program.cs

```diff
@@ -470,66 +472,123 @@ static P2pEndPoint GetExternalEndPoint(Socket socket)
                     Internal = (socket.LocalEndPoint as IPEndPoint)
                 };
             }
-
+            
             Console.WriteLine("Could not find a working STUN server");
+
             return null;
         }
 
-
         static int SleepTime(DateTime now)
         {
-            List<int> seconds = new List<int>() {10, 20, 30, 40, 50, 60};
-
             int next = seconds.Find(x => x > now.Second);
 
             return next - now.Second;
         }
 
-        static UdtSocket PeerConnect(Socket socket, string remoteAddr, int remotePort)
+        static async Task<UdtSocket> PeerConnectAsync(string remoteAddress, int[] remotePorts)
         {
-            bool bConnected = false;
-            int retry = 0;
+            if (!sockets.Any())
+                return null;
+
+            Console.WriteLine("Trying to connect to peer on different ports");
 
-            UdtSocket client = null;
+            UdtSocket udtSocket;
+            CancellationTokenSource cts;
+            var retry = 1;
 
-            while (!bConnected)
+            do
             {
-                try
-                {
-                    DateTime now = InternetTime.Get();
+                Console.WriteLine($"Try {retry++}:");
+                Console.WriteLine();
 
-                    int sleepTimeToSync = SleepTime(now);
+                cts = new CancellationTokenSource();
 
-                    Console.WriteLine("[{0}] - Waiting {1} sec to sync with other peer",
-                        now.ToLongTimeString(),
-                        sleepTimeToSync);
-                    System.Threading.Thread.Sleep(sleepTimeToSync * 1000);
+                var taskList = remotePorts.Select((port, index) => ConnectAsync(sockets[index], remoteAddress, port, cts.Token)).ToList();
 
-                    GetExternalEndPoint(socket);
+                var now = InternetTime.Get();
+                var sleepTimeToSync = SleepTime(now);
 
-                    if (client != null)
-                        client.Close();
+                Console.WriteLine("[{0}] - Waiting {1} sec to sync with other peer", now.ToLongTimeString(), sleepTimeToSync);
 
-                    client = new UdtSocket(socket.AddressFamily, socket.SocketType);
+                await Task.Delay(TimeSpan.FromSeconds(sleepTimeToSync), cts.Token);
+
+                Console.WriteLine("Please wait... ");
+                udtSocket = await (await Task.WhenAny(taskList));

```

Comment:
two consecutive awaits? this looks weird

Reply:
First await return the task and the second await gets result from the task.